src/aws/flb_aws_error_reporter.c

```diff
@@ -0,0 +1,247 @@
+
+    /* create the message Linked Lists */
+    mk_list_init(&error_reporter->messages);
+
+    return error_reporter;
+}
+
+/* error reporter write the error message into reporting file and memory*/
+int flb_aws_error_reporter_write(struct flb_aws_error_reporter *error_reporter, char *msg)
+{
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct flb_error_message *message;
+    struct flb_error_message *tmp_message;
+    flb_sds_t buf;
+    flb_sds_t buf_tmp;
+    int deleted_message_count = 0;
+    FILE *f;
+
+    buf = flb_sds_create(msg);
+    if (!buf) {
+        flb_errno();
+        return -1;
+    }
+    /* check if the message is the same with latest one in queue*/
+    if (mk_list_is_empty(&error_reporter->messages) != 0) {
+        tmp_message = mk_list_entry_last(&error_reporter->messages,
+                                         struct flb_error_message, _head);
+        if (tmp_message->len == flb_sds_len(buf) &&
+                flb_sds_cmp(tmp_message->data, buf, tmp_message->len) == 0) {
+
+            tmp_message->timestamp = time(NULL);
+            flb_sds_destroy(buf);
+            return 0;
+        }
+    }
+
+    message = flb_malloc(sizeof(struct flb_error_message));
+    if (!message) {
+        flb_sds_destroy(buf);
+        flb_errno();
+        return -1;
+    }
+    /* check if new message is too large and truncate*/
+    if (flb_sds_len(buf) > error_reporter->max_size) {
+        // truncate message
+        buf_tmp = flb_sds_copy(buf, msg, error_reporter->max_size);
+        if (buf_tmp != buf) {
+            buf = buf_tmp;
+        }
+    }
+
+    message->data = flb_sds_create(buf);
+    if (!message->data) {
+        flb_sds_destroy(buf);
+        flb_free(message);
+        return -1;
+    }
+
+    message->len = flb_sds_len(buf);
+
+   /* clean up old message to provide enough space for new message*/
+    mk_list_foreach_safe(head, tmp, &error_reporter->messages) {
+        tmp_message = mk_list_entry(head, struct flb_error_message, _head);
+        if (error_reporter->file_size + flb_sds_len(buf) <= error_reporter->max_size) {
+            break;
+        }
+        else {
+            error_reporter->file_size -= tmp_message->len;
+            deleted_message_count++;
+            mk_list_del(&tmp_message->_head);
+            flb_sds_destroy(tmp_message->data);
+            flb_free(tmp_message);
+        }
+    }
+    message->timestamp = time(NULL);
+
+    mk_list_add(&message->_head, &error_reporter->messages);
+
+    if (deleted_message_count == 0) {
+        f = fopen(error_reporter->file_path, "a");
+        fprintf(f, message->data);
+    }
+    else {
+        f = fopen(error_reporter->file_path, "w");
+        mk_list_foreach_safe(head, tmp, &error_reporter->messages) {
+            tmp_message = mk_list_entry(head, struct flb_error_message, _head);
+            fprintf(f, tmp_message->data);
+        }
+    }
+    fclose(f);
+
+    error_reporter->file_size += message->len;
```

Comment:
This call should be moved up a few lines to where you add teh new message to the list. This is better for readability and maintainability.

Reply:
moved up as suggestion